### ISSUE TITLE
DPMSolverMultistepScheduler has wrong number of timesteps given num_inference_steps with StableDiffusionXL pipelines

### ISSUE THREAD

**Issue by AmericanPresidentJimmyCarter** [huggingface/diffusers#4122]
### Describe the bug

For example when 100 steps are given for num_inference_steps, only 95 are performed.

### Reproduction

```py
from diffusers import StableDiffusionXLPipeline
from diffusers.schedulers import DPMSolverMultistepScheduler
import torch

scheduler = DPMSolverMultistepScheduler(
    beta_start=0.00085,
    beta_end=0.012,
    beta_schedule="scaled_linear",
    prediction_type="epsilon",
    num_train_timesteps=1000,
    trained_betas=None,
    thresholding=False,
    algorithm_type="dpmsolver++",
    solver_type="midpoint",
    lower_order_final=True,
    use_karras_sigmas=True,
)

hf_cache_dir = '/path/to/hf'

pipe = StableDiffusionXLPipeline.from_pretrained(
    f"{hf_cache_dir}/stable-diffusion-xl-base-0.9", variant="fp16", use_safetensors=True,
    torch_dtype=torch.bfloat16,
)
pipe.enable_model_cpu_offload(0)
pipe.enable_vae_slicing()
pipe.enable_vae_tiling()
pipe.register_modules(scheduler=scheduler)

with torch.no_grad(), torch.amp.autocast('cuda', dtype=torch.bfloat16):
    random_generator = torch.Generator()
    random_generator.manual_seed(12345)
    prompt = "Astronaut in a jungle, cold color palette, muted colors, detailed, 8k"
    image = pipe(prompt=prompt, output_type="pil",
        num_inference_steps=100, generator=random_generator).images[0]
```

### Logs

```shell
python test_sdxl2.py 
Loading pipeline components...: 100%|███████████████████████████████████████████████████████████████████| 7/7 [00:02<00:00,  2.81it/s]
Loading pipeline components...: 100%|███████████████████████████████████████████████████████████████████| 5/5 [00:03<00:00,  1.65it/s]
 95%|███████████████████████████████████████████████████████████████████████████████████████████▏    | 95/100 [00:32<00:01,  2.91it/s]
```

### System Info

Commit b024ebb96501b410ff21c8158cd075ceb3856404 (main)

### Who can help?

@patrickvonplaten @wil

**Comment by patrickvonplaten** [comment]
That's a great catch! Thanks a lot for the reproducible

**Comment by patrickvonplaten** [comment]
@yiyixuxu @sayakpaul anybody time to take this one?

**Comment by yiyixuxu** [comment]
@patrickvonplaten 
so it is the similar situation we had with the heun scheduler in shap-e PR - we ended up duplicated timesteps when `use_karras_sigmas=True` and it got deduplicated here

time to add a `index_counter`? 

https://github.com/huggingface/diffusers/blob/main/src/diffusers/schedulers/scheduling_dpmsolver_multistep.py#L279

**Comment by patrickvonplaten** [comment]
Nice find! Yes I think it's time for the index_counter in this case, would be great if you could open a PR :-)

**Comment by yangdeai** [comment]
I met the same problem too, it always miss last one step. Is it fixed?

```

  kwargs = {"use_karras_sigmas": True}
  pipe.scheduler = DPMSolverMultistepScheduler.from_config(pipe.scheduler.config, **kwargs)

```

![image](https://github.com/huggingface/diffusers/assets/55951120/904a9cbe-2850-49a3-bca8-bea46455c053)

**Comment by yiyixuxu** [comment]
@yangdeai 
should be fixed with this PR merged now https://github.com/huggingface/diffusers/pull/4986

**Comment by yangdeai** [comment]
> @yangdeai should be fixed with this PR merged now #4986

Great! Can I use the fixed code just install diffusers with: pip install diffusers?
